projects/slap_manipulation/src/slap_manipulation/agents/slap_agent.py

```diff
@@ -0,0 +1,333 @@
+import datetime
+import os
+from typing import Any, Dict
+
+import numpy as np
+import torch
+import trimesh
+import yaml
+from slap_manipulation.policy.action_prediction_module import ActionPredictionModule
+from slap_manipulation.policy.interaction_prediction_module import (
+    InteractionPredictionModule,
+)
+from slap_manipulation.utils.data_processing import (
+    filter_and_remove_duplicate_points,
+    voxelize_point_cloud,
+)
+from slap_manipulation.utils.pointcloud_preprocessing import (
+    get_local_action_prediction_problem,
+)
+
+from home_robot.utils.point_cloud import show_point_cloud
+from home_robot_hw.env.stretch_abstract_env import GRIPPER_IDX
+
+
+class SLAPAgent(object):
+    """
+    Combined E2E agent which uses the SLAP architecture
+    to predict actions given a language instruction
+    """
+
+    def __init__(self, cfg, device="cuda", task_id: int = -1):
+        self.task_id = task_id
+        print("[SLAPAgent]: task_id = ", task_id)
+        self._dry_run = cfg.SLAP.dry_run
+        self.interaction_point = None
+        self.cfg = cfg
+        self.device = device
+        self._curr_keyframe = -1
+        self._last_action = None
+        self._min_depth = self.cfg.SLAP.min_depth
+        self._max_depth = self.cfg.SLAP.max_depth
+        self._feat_dim = 1
+        # self._x_max = self.cfg.SLAP.y_max
+        self._z_min = self.cfg.SLAP.z_min
+        self._voxel_size_1 = self.cfg.SLAP.voxel_size_1
+        self._voxel_size_2 = self.cfg.SLAP.voxel_size_2
+        if cfg.SLAP.APM.skill_to_action_file is not None:
+            self.skill_to_action = yaml.load(
+                open(cfg.SLAP.APM.skill_to_action_file, "r"), Loader=yaml.FullLoader
+            )
+        else:
+            self.skill_to_action = None
+
+    def get_goal_info(self):
+        info = {}
+        info["task-name"] = self.cfg.EVAL.task_name
+        info["object_list"] = self.cfg.EVAL.object_list
+        info["num-actions"] = self.cfg.EVAL.num_keypoints
+        return info
+
+    def load_models(self):
+        self.interaction_prediction_module = InteractionPredictionModule()
+        self.action_prediction_module = ActionPredictionModule(self.cfg.SLAP.APM)
+        self.interaction_prediction_module.load_weights(self.cfg.SLAP.IPM.path)
+        self.interaction_prediction_module.to(self.device)
+        self.action_prediction_module.load_weights(self.cfg.SLAP.APM.path)
+        self.action_prediction_module.to(self.device)
+        print("Loaded SLAP weights")
+
+    def get_proprio(self):
+        # if self._last_action is None:
+        return np.array([2, 2, 2, 2, 2, 2, 2, -1])
```

Comment:
why always the same?

Reply:
It's what we initialize proprio as for the LSTM. This is updated internally in APM per inference.